src/List.mo

```diff
@@ -1534,18 +1704,35 @@ module {
   ///
   /// *Runtime and space assumes that `compare` runs in O(1) time and space.
   public func max<T>(list : List<T>, compare : (T, T) -> Order.Order) : ?T {
-    if (isEmpty(list)) return null;
-
-    var maxSoFar = at(list, 0);
-    forEach<T>(
-      list,
-      func(x) = switch (compare(x, maxSoFar)) {
-        case (#greater) maxSoFar := x;
-        case _ {}
-      }
-    );
+    var maxSoFar : T = switch (first(list)) {
+      case (?x) x;
+      case null return null
+    };
 
-    return ?maxSoFar
+    let blocks = list.blocks;
+    let blockCount = blocks.size();
+
+    var i = 1;
+    while (i < blockCount) {
+      let db = blocks[i];
+      let sz = db.size();
+      if (sz == 0) return ?maxSoFar;
+
+      var j = 0;
+      while (j < sz) {
+        switch (db[j]) {
+          case (?x) switch (compare(x, maxSoFar)) {
+            case (#greater) maxSoFar := x;
+            case _ {}
+          };
+          case null return ?maxSoFar
+        };
+        j += 1
+      };
+      i += 1
+    };
+
+    ?maxSoFar
```

Comment:
Are we comparing here the first element against itself? Can we start the loop with the second element?